packages/babel-parser/src/util/location.js

```diff
@@ -12,10 +10,14 @@ export type Pos = {
 export class Position {
   line: number;
   column: number;
+  index: number;
 
-  constructor(line: number, col: number) {
+  constructor(line: number, col: number, index: number) {
     this.line = line;
     this.column = col;
+
+    // this.index = index;
+    Object.defineProperty(this, "index", { enumerable: false, value: index });
```

Comment:
This should probably go in a minor, because it's a new observable property on `loc` objects in the AST (making it non-enumerable hides it from the fixtures, but it's still there :stuck_out_tongue:).

However, I would like to merge this PR as soon as possible because:
1. This PR also fixes different `errorRecovery` exceptions and some other bugs
2. This PR conflicts with virtually everything touching the parser, and rebasing it will not be easy.

For now, we can hide this index property in a new `const indexes = new WeakMap()` exported by this file, and do
```suggestion
    // this.index = index;
    indexes.set(this, index);
```

And replace `/([\w\.]+)\.index/g` with `indexes.get($1)`. Then, we can discuss if we want to expose it as a public property in Babel 7.17.0.

Reply:
Sounds good, can make that change.

Reply:
@nicolo-ribaudo @tolmasky despite swap to `WeakMap` just having typed member in https://github.com/babel/babel/blob/b05dad7fed07672514fa6d0d21ce4c1e2c3a6f79/packages/babel-parser/src/util/location.js#L13

seems to actually init this field with `void 0` ( see https://www.runpkg.com/?@babel/parser@7.16.10/lib/index.js#9 ).

This did trip some of our tests ( TBF the test was in practice to strict and didn't allow for any uncounted for fields, so we will be relaxing them a bit ), but most importantly, I just want to check with you wether this is acceptable for you given the comments above

Reply:
Hi @pieh, I actually reported this immediately after we shipped as I also noticed it: https://github.com/babel/babel/issues/14182. Apologies though, as I had the discussion with @nicolo-ribaudo offline, where he advised that it was probably better to just hash it out here: https://github.com/babel/babel/pull/14174, as we might just be adding them in fully in the next minor update. But I'll let @nicolo-ribaudo comment more here as I am definitely open to removing that line for the next patch update if its causing any issues.

Reply:
There is no problem here from my side (as I mentioned - our tests/validation were just too strict and brittle because of that - I really should just validate keys we do need (which is `line`/`column`, but don't enforce that they are only possible fields in `Position`). This is especially true as `babel` is not the only "source" we are handling errors from. 

My comment was just to make sure you are aware as this sounds like something you wanted to avoid in patch release. For us it really doesn't matter (in practice, and not semantics) if it's a patch or minor given that right now we do use `^` version selector (and not pinned version or `~` for pinned minor)

Reply:
Sure thing, and again, apologies! This one's on me, I just forgot to delete the declaration after switching the uses of it.